DAG-681: Add Energy Statistic Homogeneity Calculator

anjani-bhat opened the pull request.

ancostas commented on src/signal_processing_algorithms/homogeneity.py:
```diff
@@ -0,0 +1,93 @@
+
+        :param avg: Function to calculate average.
+        """
+        self._avg = avg
```
Considering that this can probably be removed, I think we don't even need a class here.

ancostas commented on src/signal_processing_algorithms/homogeneity.py:
```diff
@@ -0,0 +1,93 @@
+
+        common_term = 2 * expectation_x_y
+        inhomogeneity_coefficient = (common_term - expectation_x - expectation_y) / common_term
+        assert 0 <= inhomogeneity_coefficient <= 1
```
Assert in source code? I think you want to raise your own error, or do an assert in a test instead. Probably the latter since this just checks that your calculation worked.

ancostas commented on src/signal_processing_algorithms/difference.py:
```diff
@@ -0,0 +1,18 @@
+    """
+    if y is None:
+        y = x
+    diff = np.abs(x[:, None] - y)
```
is `np.abs` always the euclidean distance between two data points? A minor change here will allow us to handle differences between vectors as well.

ancostas commented on src/signal_processing_algorithms/difference.py:
```diff
@@ -0,0 +1,18 @@
+import numpy as np
+
+
+def pairwise_difference(x: np.ndarray, y: np.ndarray = None) -> np.ndarray:
```
These are pairwise differences, not just one difference, right?  Or perhaps a distance matrix.

ancostas reviewed (changes requested):
Good work, I think we should double down on the results you've had here so far and go a bit further. Will shoot you an invite to discuss.

anjani-bhat commented:
This is partially done: It includes all the changes for multivariate Energy Statistics calculations. 
This does not include the changes to E-Divisive Means to make it multivariate.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+
+
+@dataclass
+class EnergyStatisticsTestResult:
```
I guess technically this is `EnergyStatistics` and the other is `EnergyStatisticsCalculator` or something like that.

anjani-bhat replied:
Made these two classes instead since EnergyCalculator should ideally not have e, t, h as its attributes and the pvalues are not exactly energy statistics but results of permutation tests.
  ```
 class EnergyStatistics:
    e: float
    t: float
    h: float
```

  ```
class EnergyStatisticsWithProbabilities(EnergyStatistics):
    e_pvalue: float
    t_pvalue: float
    h_pvalue: float
```

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+        )
+
+    @staticmethod
+    def get_t_from_e(t_coefficient: float, e_statistic: float) -> float:
```
Do we need a method to multiply two values together?

anjani-bhat replied:
Added this as function to avoid having to reuse same calculations. Have removed this now as its more readable to do calculations without using these funcitons.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+
+        # Test statistic
+        self._t_coefficient = (self._n * self._m) / (self._n + self._m)
+        self.t = self.get_t_from_e(self._t_coefficient, self.e)
```
Probably clearer to just combine these in to one line and remove `get_t_from_e`.

anjani-bhat replied:
Done.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+        return f"EnergyStatistics(e:{self.e}, t:{self.t}, h:{self.h})"
+
+    @staticmethod
+    def get_e_from_distance_matrices(
```
We also probably don't need a method for this kind of calculation -- it is clearer in plaintext.

anjani-bhat replied:
Done.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+        return t_coefficient * e_statistic
+
+    @staticmethod
+    def get_h_from_e_and_distance_matrix(e: float, distances_between_xy: np.ndarray) -> float:
```
Same comment as other two methods.

anjani-bhat replied:
Done.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+
+        distances_within_x = get_distance_matrix(self._x, self._x)
+        distances_within_y = get_distance_matrix(self._y, self._y)
+        distances_between_xy = get_distance_matrix(self._x, self._y)
```
In your other methods you are calculating `np.mean(distances_between_xy)` more than once. We can calculate the expected values once and then just combine them in different ways to get e, t, h in a clear way.

anjani-bhat replied:
Done.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+    e: float
+    t: float
+    h: float
+    e_p: float
```
pvalue rather than p would make this clearer for the next three things.

anjani-bhat replied:
Done.

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+            # np.ix_()'s main use is to create an open mesh so that we can use it to
+            # select specific indices from an array (specific sub-array).
+            # We use it here to rearrange the combined distance matrix.
+            shuffled_distances = distances_between_all[np.ix_(row_indices, row_indices)]
```
You deserve a gold star for this.

anjani-bhat replied:
Thank you !!

ancostas commented on src/signal_processing_algorithms/energy_statistics.py:
```diff
@@ -0,0 +1,309 @@
+            distances_within_x, distances_within_y, distances_between_xy
+        )
+
+    def get_statistics_and_probabilities(self, num_tests: int = 1000) -> EnergyStatisticsTestResult:
```
This is looking good, but I think it will clean up the design/APIs a bit to separate concerns:
* difference matrix calculation (this is already separate)
* energy statistics calculation (based on x,y OR diff_matrix, n, m)
* permutation testing (is just shuffling stuff and otherwise using the above two for calculations)

anjani-bhat replied:
Done.

ancostas reviewed (changes requested):
Some cleanup, but looking good.

ancostas commented on src/signal_processing_algorithms/energy_statistics/energy_statistics.py:
```diff
@@ -0,0 +1,366 @@
+    # sum |Xi - Xj| for i < j < tau
+    x = 0
+    # sum |Yi - Yj| for tau <= i < j
+    y = np.sum(diffs)
```
Technically we could make this still do just half of the diff matrix and then double the terms due to symmetry arguments. We can probably leave that for a future optimization ticket.

anjani-bhat replied:
Made the optimization

ancostas reviewed (changes requested):
I think I broke your tests, and the docs need an update at this point.

ancostas commented on src/signal_processing_algorithms/energy_statistics/e_divisive.c:
```diff
@@ -113,19 +113,20 @@ bool qhat_values(double * diffs, double * qhat_values, int length){
     int tau;
     for(tau=0;tau<length;tau++) {
-        qhat_values[tau] = calculate_q(cross_term, x_term, y_term, tau, length - tau);
+        t_stat_values[tau] = calculate_t(cross_term, 2*x_term, 2*y_term, tau, length - tau);
```
Why are we doubling the between terms, but not the cross term?

anjani-bhat replied:
x = [[1, 2], [3, 4], [5, 6]]
    y = [[7, 8], [9, 10]]

    Combined matrix is:
    [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]]
    ----------------------------------------------
    Pairwise distances within Combined matrix is:
    ----------------------------------------------
    [
        [ 0.          2.82842712  5.65685425  8.48528137 11.3137085 ]
        [ 2.82842712  0.          2.82842712  5.65685425  8.48528137]
        [ 5.65685425  2.82842712  0.          2.82842712  5.65685425]
        [ 8.48528137  5.65685425  2.82842712  0.          2.82842712]
        [11.3137085   8.48528137  5.65685425  2.82842712  0.        ]
    ]
```
The distance matrix of x is:
[[0.        , 2.82842712, 5.65685425],
 [2.82842712, 0.        , 2.82842712],
 [5.65685425, 2.82842712, 0.        ]]

The distance matrix of y is:
[[0.        , 2.82842712],
 [2.82842712, 0.        ]]

Both of which are symmetric - so we can calculate the sum of just either upper or lower diagonal and double it.

However the distance matrix between x and y is:
[[ 8.48528137, 11.3137085 ],
  [ 5.65685425,  8.48528137],
  [ 2.82842712,  5.65685425]]
which is either the bottom left or top right part of the combined matrix above.

Since it is not symmetric and doesn't need to be accounted twice (there is no "distance from itself" concept here and its just sum of xi-yj), we only need to add the cross term once.
i.e we can use the same cross term that we used for in the earlier implementation of e-divisive (only the way how we calculate pairwise distances WITHIN x and pairwise distances WITHIN y has changed, the pairwise distance BETWEEN the x and y remains same as before).

We can jump on a call if it's not clear.
```

ancostas replied:
Well, the cross term is symmetric over the diagonal (transpose symmetry), but I think the valid point you're making is that we shouldn't be calculating over the whole matrix since to begin with since that would be 2*n*m entries which we then average over n*m --> incorrect!

ancostas commented on src/signal_processing_algorithms/energy_statistics/energy_statistics.py:
```diff
@@ -0,0 +1,408 @@
+        # Since the distance matrix is symmetric,
+        # we can consider the upper or lower diagonal triangle and double the sum for
+        # calculating the pairwise distances within x as well as y.
+        statistics[tau] = _calculate_stats(2 * x, 2 * y, xy, tau, len(distance_matrix) - tau)[1]
```
Same comment, why doubling within distances, but not between? Seems like if we are only taking the sum of the upper triangle, we should double all terms to equal the sum of the whole matrix due to symmetry.

ancostas reviewed (commented):
Looking good. All that remains is the investigation around random shuffle usage and whether that is what caused differences in numbers of change points detected.

anjani-bhat commented:
Observations upon investigation of different change points:
1. There are tests currently in master (`test_detection` > `test_large`, `test_very_large` and `test_huge`) that are incorrect. These are the ones with the `@pytest.mark.slow` and don’t get run unless explicitly specified and probably that is why it was never caught before merging in the code. ( `test_large` produces 34 cps instead of 31 specified in the test file, `test_very_large` produces 69 instead of 61 as specified in the test file, `test_huge` produces 138 instead of 132 as specified in the test file)
2. `np.random.shuffle` and `random.shuffle` produces different shuffle order for the same seed.
3. To validate 2, changed `np.random.shuffle` to `random.shuffle` in the new implementation and ran the it - but the results observed were not the same for large timeseries. 
4. To further investigate 3, changed the `qhat` formula to t formula in old implementation and ran again to obtain the same results. The order of the change points found for really large time series becomes different from old implementation. For huge time series the number of change points found varies as well.
So we can conclude that the changes in the number of change points detected are due to: 1 and 2 above and the change from `qhat` to `t-stat` formula.

If we want the same change points to be generated as the older implementation for MOST of the time series (cannot guarantee this for all time series because of change made from `qhat` to `t-stat`. The changes are mostly seen in large time series.):
1. We can replace `np.random.shuffle` with `random.shuffle`
2. Use the `seed` 1234 for change point calculation in edivisive.

anjani-bhat commented:
An example for this change:

For qhat based calculation - the change points observed for `very_large.json`:
[2364, 2318, 25, 64, 114, 173, 198, 237, 287, 346, 371, 410, 460, 519, 544, 583, 633, 692, 717, 756, 806, 865, 890, 929, 979, 1038, 1063, 1102, 1152, 1211, 1236, 2190, 2145, 2249, 2274, 2017, 1972, 2076, 2101, 1844, 1799, 1903, 1928, 1671, 1626, 1730, 1755, 1498, 1453, 1557, 1582, 1325, 1279, 1384, 1409, 69, 242, 415, 588, 761, 934, 1107, 1448, 1621, 1794, 1967, 2140, 2313, 1275]
For t-stat based calculation this changed to:
[2364, 2318, 25, 2190, 2145, 2249, 2274, 2017, 1972, 2076, 2101, 1844, 1799, 1903, 1928, 1671, 1626, 1730, 1755, 1498, 1453, 1557, 1582, 1325, 1280, 1384, 1409, 1152, 1107, 1211, 1236, 979, 934, 1038, 1063, 806, 761, 865, 890, 633, 588, 692, 717, 460, 415, 519, 544, 287, 242, 346, 371, 114, 68, 173, 198, 237, 410, 583, 756, 929, 1102, 1275, 1448, 1621, 1794, 1967, 2140, 2313]

The length reduced from 69 to 68 and the order of the change points also changed as above.

anjani-bhat commented:
evergreen merge
